File: day17.go
package main

import (
	"fmt"
	"time"
)

func main() {

	// Lets create a buffered channel with capacity 2.
	// We can send 2 items without blocking.
	ch := make(chan string, 2)
	ch <- "Message 1"
	ch <- "Message 2"

	fmt.Println("Sent 2 messages without a receiver yet")

	fmt.Println("Received:", <-ch)

	// Closing and ranging
	numbers := make(chan int)

	go func() {
		for i := 1; i <= 5; i++ {
			numbers <- i
			time.Sleep(200 * time.Millisecond)
		}

		// IMP: WE must close the channel when done.
		close(numbers)
		fmt.Println("Channel closed.")
	}()

	fmt.Println("Waiting for numbers...")
	// range loops over the channel.
	// it automatically receives values.
	// automatically stops when channel is empty and close.
	for num := range numbers {
		fmt.Println("Received number is :", num)
	}

	fmt.Println("Done.")
}

File: python/day17.py
import queue
import threading
import time

def producer(q):
    for i in range(1,6):
        q.put(i)
        time.sleep(0.2)

    # Simulate closing by sending a none value
    q.put(None)
    print("Producer finished")

    #  main execution
    # create a buffered queue
    q = queue.Queue(maxsize=2)
    q.put("Message 1")
    q.put("Message 2")

    print("Sent 2 message without a receiver yet")
    print(f"Received :{q.get()}")

# looping

num_q = queue.Queue()
t = threading.Thread(target=producer, args=(num_q,))
t.start()

print("waiting for numbers")

while True:
    val = num_q.get()
    if val is None:
        break
    print(f"Received number: {val}")

print("Done")